device: move {In,Out}boundElement Mutex to container type

jwhited opened the pull request:
{In,Out}boundElement locking can contribute to significant overhead via sync.Mutex.lockSlow() in some environments. These types are passed throughout the device package as elements in a slice, so move the per-element Mutex to a container around the slice.

bradfitz commented on device/uapi.go:
```diff
@@ -420,7 +420,14 @@ func (device *Device) IpcHandle(socket net.Conn) {
 	case "set=1\n":
-		status = device.IpcSetOperation(buffered.Reader)
+		err := device.IpcSetOperation(buffered.Reader)
+		if err != nil {
```
technically no need for this line but also fine. I guess it's explicit.

bradfitz commented on device/uapi.go:
```diff
@@ -105,7 +105,7 @@ func (device *Device) IpcGetOperation(socket *bufio.Writer) *IPCError {
 }
 
-func (device *Device) IpcSetOperation(socket *bufio.Reader) *IPCError {
+func (device *Device) IpcSetOperation(socket *bufio.Reader) error {
```
I'd change the commit message to say "not return typed error". I assumed when reading it that "return an error" meant you went from returning nothing to returning an error.

jwhited committed:
device: move Queue{In,Out}boundElement Mutex to container type

Queue{In,Out}boundElement locking can contribute to significant
overhead via sync.Mutex.lockSlow() in some environments. These types
are passed throughout the device package as elements in a slice, so
move the per-element Mutex to a container around the slice.

Signed-off-by: Jordan Whited <jordan@tailscale.com>

valscale reviewed (commented):
This is the first time I've looked at this code, but it looks good to me!

maisem commented on device/pools.go:
```diff
@@ -65,28 +65,32 @@ func (device *Device) PopulatePools() {
-	return device.pool.outboundElementsSlice.Get().(*[]*QueueOutboundElement)
+func (device *Device) GetInboundElementsContainer() *QueueInboundElementsContainer {
+	c := device.pool.inboundElementsContainer.Get().(*QueueInboundElementsContainer)
+	c.Mutex = sync.Mutex{}
```
This looks sus, what're you trying to do?

maisem replied:
I think I've convinced myself that the locking is just plain weird here, and that you aren't making it worse.

maisem commented on device/send.go:
```diff
@@ -46,18 +46,21 @@ import (
 func (device *Device) NewOutboundElement() *QueueOutboundElement {
 	elem := device.GetOutboundElement()
 	elem.buffer = device.GetMessageBuffer()
-	elem.Mutex = sync.Mutex{}
```
Oh this is so weird

maisem commented on device/pools.go:
```diff
@@ -46,13 +46,13 @@ func (p *WaitPool) Put(x any) {
-	device.pool.inboundElementsSlice = NewWaitPool(PreallocatedBuffersPerPool, func() any {
+	device.pool.inboundElementsContainer = NewWaitPool(PreallocatedBuffersPerPool, func() any {
 		s := make([]*QueueInboundElement, 0, device.BatchSize())
-		return &s
+		return &QueueInboundElementsContainer{elems: s}
```
you can inline this now
